include/aos/common/resourcemonitor.hpp

```diff
@@ -0,0 +1,301 @@
+     * @return Error
+     */
+    virtual Error GetInstanceMonitoringData(const String& instanceID, MonitoringData& monitoringData) = 0;
+};
+
+/**
+ * Monitor sender interface.
+ */
+class SenderItf {
+public:
+    /**
+     * Sends monitoring data
+     *
+     * @param monitoringData monitoring data
+     * @return Error
+     */
+    virtual Error SendMonitoringData(const NodeMonitoringData& monitoringData) = 0;
+};
+
+/**
+ * Resource monitor interface.
+ */
+class ResourceMonitorItf {
+public:
+    /**
+     * Destructor
+     */
+    virtual ~ResourceMonitorItf() = default;
+
+    /**
+     * Gets the node info object
+     *
+     * @param nodeInfo node info
+     * @return Error
+     */
+    virtual Error GetNodeInfo(NodeInfo& nodeInfo) const = 0;
+
+    /**
+     * Starts instance monitoring
+     *
+     * @param instanceID instance ident
+     * @param monitoringConfig monitoring config
+     * @return Error
+     */
+    virtual Error StartInstanceMonitoring(const String& instanceID, const InstanceMonitorParams& monitoringConfig) = 0;
+
+    /**
+     * Stops instance monitoring
+     *
+     * @param instanceID instance ident
+     * @return Error
+     */
+    virtual Error StopInstanceMonitoring(const String& instanceID) = 0;
+};
+
+/**
+ * Resource monitor.
+ */
+class ResourceMonitor : public ResourceMonitorItf, public SubscriberItf {
+public:
+    /**
+     * Constructor
+     */
+    ResourceMonitor() = default;
+
+    /**
+     * Destructor
+     */
+    ~ResourceMonitor();
+
+    /**
+     * Initializes resource monitor
+     *
+     * @param resourceUsageProvider resource usage provider
+     * @param monitorSender monitor sender
+     * @return Error
+     */
+    Error Init(
+        ResourceUsageProviderItf& resourceUsageProvider, SenderItf& monitorSender, aos::SubscribableItf& subscribable);
+
+    /**
+     * Gets the node info object
+     *
+     * @param nodeInfo node info
+     * @return Error
+     */
+    Error GetNodeInfo(NodeInfo& nodeInfo) const override;
+
+    /**
+     * Starts instance monitoring
+     *
+     * @param instanceID instance ident
+     * @param monitoringConfig monitoring config
+     * @return Error
+     */
+    Error StartInstanceMonitoring(const String& instanceID, const InstanceMonitorParams& monitoringConfig) override;
+
+    /**
+     * Stops instance monitoring
+     *
+     * @param instanceID instance ident
+     * @return Error
+     */
+    Error StopInstanceMonitoring(const String& instanceID) override;
+
+    /**
+     * Respond to a connection event.
+     */
+    void OnConnect() override;
+
+    /**
+     * Respond to a disconnection event.
+     */
+    void OnDisconnect() override;
+
+private:
+    static constexpr auto cTimeoutSend = AOS_CONFIG_MONITORING_SEND_PERIOD_SEC;
+
+    ResourceUsageProviderItf* mResourceUsageProvider {};
+    SenderItf*                mMonitorSender {};
+    aos::SubscribableItf*     mSubscribable {};
```

Comment:
this itf not defined yet, remove from this commit.

Reply:
Done